radix-engine-stores/src/hash_tree/tree_store.rs

```diff
@@ -56,80 +59,162 @@ pub trait ReadableTreeStore {
-    /// A direct constructor.
-    pub fn new(base: &'s R) -> StagedTreeStore<'s, R> {
-        StagedTreeStore {
-            base,
-            overlay: MemoryTreeStore::new(),
+impl SerializedInMemoryTreeStore {
+    /// A constructor of a newly-initialized, empty store.
+    pub fn new() -> Self {
+        Self {
+            memory: HashMap::new(),
+            stale_key_buffer: Vec::new(),
         }
     }
 }
 
-impl<'s, R: ReadableTreeStore> ReadableTreeStore for StagedTreeStore<'s, R> {
+impl ReadableTreeStore for SerializedInMemoryTreeStore {
     fn get_node(&self, key: &NodeKey) -> Option<TreeNode> {
-        self.overlay
-            .get_node(key)
-            .or_else(|| self.base.get_node(key))
+        self.memory
+            .get(&encode_key(key))
+            .map(|bytes| scrypto_decode(bytes).unwrap())
+    }
+}
+
+impl WriteableTreeStore for SerializedInMemoryTreeStore {
+    fn insert_node(&mut self, key: NodeKey, node: TreeNode) {
+        self.memory
+            .insert(encode_key(&key), scrypto_encode(&node).unwrap());
+    }
+
+    fn record_stale_node(&mut self, key: NodeKey) {
+        self.stale_key_buffer.push(encode_key(&key));
+    }
+}
+
+/// Encodes the given node key in a format friendly to Level-like databases (i.e. strictly ordered
+/// by numeric version).
+pub fn encode_key(key: &NodeKey) -> Vec<u8> {
+    let m = &key.version().to_be_bytes();
+    let x = &[(key.nibble_path().num_nibbles() % 2) as u8; 1];
+    let k = key.nibble_path().bytes();
+    [m, k, x].concat()
+}
+
+// Note: We need completely custom serialization scheme only for the node keys. The remaining
+// structures can simply use SBOR, with only the most efficiency-sensitive parts having custom
+// codecs, implemented below:
+
+impl Categorize<ScryptoCustomValueKind> for Nibble {
+    #[inline]
+    fn value_kind() -> ValueKind<ScryptoCustomValueKind> {
+        ValueKind::U8
     }
 }
 
-impl<'s, R: ReadableTreeStore> WriteableTreeStore for StagedTreeStore<'s, R> {
-    fn insert_node(&mut self, key: &NodeKey, node: TreeNode) {
-        self.overlay.insert_node(key, node);
+impl<E: Encoder<ScryptoCustomValueKind>> Encode<ScryptoCustomValueKind, E> for Nibble {
+    #[inline]
+    fn encode_value_kind(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        encoder.write_value_kind(Self::value_kind())
     }
 
-    fn record_stale_node(&mut self, key: &NodeKey) {
-        self.overlay.record_stale_node(key)
+    #[inline]
+    fn encode_body(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        encoder.write_byte(u8::from(*self))?;
+        Ok(())
+    }
+}
+
+impl<D: Decoder<ScryptoCustomValueKind>> Decode<ScryptoCustomValueKind, D> for Nibble {
+    fn decode_body_with_value_kind(
+        decoder: &mut D,
+        value_kind: ValueKind<ScryptoCustomValueKind>,
+    ) -> Result<Self, DecodeError> {
+        decoder.check_preloaded_value_kind(value_kind, Self::value_kind())?;
```

Comment:
Similarly, these two lines could be replaced with `Ok(Nibble(u8::decode_body_with_value_kind(decoder, value_kind)?))`